abnet3/trainer.py

```diff
@@ -256,6 +268,90 @@ def optimize_model(self, do_training=True):
+            adversarial loss will be 1 - weight
+        """
+        self.default_loss_weight = default_loss_weight
+        super().__init__(*args, **kwargs)
+
+    def optimize_model(self, do_training=True):
+        """
+        Optimization model step for the Siamese network.
+        """
+
+        train_losses = defaultdict(int)
+        dev_losses = defaultdict(int)
+        total_train_loss = 0.0
+        total_dev_loss = 0.0
+        num_batches_train = 0
+        num_batches_dev = 0
+        self.network.train()
+        for minibatch in self.dataloader.batch_iterator(train_mode=True):
+            losses = self.give_batch_to_network(minibatch)
+            train_loss_value = (self.default_loss_weight * losses['default'] +
+                                (1 - self.default_loss_weight) * losses['adversarial'])
+            self.optimizer.zero_grad()
+            if do_training:
+                train_loss_value.backward()
+                self.optimizer.step()
+            num_batches_train += 1
+            for l in losses:
+                train_losses[l] += losses[l].data[0]
+            total_train_loss += train_loss_value.data[0]
+
+        self.network.eval()
+        for minibatch in self.dataloader.batch_iterator(train_mode=False):
+            num_batches_dev += 1
+            losses = self.give_batch_to_network(minibatch)
+            for l in losses:
+                dev_losses[l] += losses[l].data[0]
+            total_dev_loss += sum(losses.values()).data[0]
+
+        normalized_train_loss = total_train_loss/num_batches_train
+        normalized_dev_loss = total_dev_loss/num_batches_dev
+        self.train_losses.append(normalized_train_loss)
+        self.dev_losses.append(normalized_dev_loss)
+        for l in train_losses:
+            train_losses[l] /= num_batches_train
+        for l in dev_losses:
+            dev_losses[l] /= num_batches_train
+
+        self.pretty_print_dict_losses("training", normalized_train_loss, train_losses)
+        self.pretty_print_dict_losses("dev", normalized_dev_loss, dev_losses)
+        return dev_losses['default']
+
+    def give_batch_to_network(self, batch):
+        X_batch1, X_batch2, y_spk_batch, y_phn_batch = batch
+        y_spk_batch = y_spk_batch.type(torch.FloatTensor)[:, None]
+        if self.cuda:
+            X_batch1 = X_batch1.cuda()
+            X_batch2 = X_batch2.cuda()
+            y_spk_batch = y_spk_batch.cuda()
+            y_phn_batch = y_phn_batch.cuda()
+        emb1_batch, emb2_batch = self.network(X_batch1, X_batch2)
+        train_loss_value = self.loss(emb1_batch, emb2_batch, y_phn_batch)
+
+        # adversarial loss
+        y_spk_predictions = self.network.run_adversarial_classifier(
+            emb1_batch, emb2_batch)
+        adversarial_loss = nn.BCELoss(size_average=self.loss.avg)(y_spk_predictions, y_spk_batch)
+        return {'default': train_loss_value, 'adversarial': adversarial_loss}
```

Comment:
Maybe schedule training batches with batches for spk adv and batch for phn discrimination.

advice from here : https://github.com/soumith/ganhacks#4-batchnorm